rust-runtime/aws-smithy-http-server/src/operation/mod.rs

```diff
@@ -0,0 +1,268 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//! # Operations
+//!
+//! The shape of a [Smithy operation] is modelled by the [`OperationShape`] trait, it's associated types
+//! [`OperationShape::Input`], [`OperationShape::Output`], and [`OperationShape::Error`] map to the structures
+//! representing the Smithy inputs, outputs, and errors respectively. When an operation error is not specified
+//! [`OperationShape::Error`] is [`Infallible`](std::convert::Infallible).
+//!
+//! A ZST for each Smithy operation is generated and [`OperationShape`] should be implemented on it, this will be used
+//! as a helper which provides static methods and parameterizes other traits.
+//!
+//! The model
+//!
+//! ```smithy
+//! operation GetShopping {
+//!    input: CartIdentifier,
+//!    output: ShoppingCart,
+//!    errors: [...]
+//! }
+//! ```
+//!
+//! is identified with the implementation
+//!
+//! ```rust
+//! # use aws_smithy_http_server::operation::OperationShape;
+//! # pub struct CartIdentifier;
+//! # pub struct ShoppingCart;
+//! # pub enum GetShoppingError {}
+//! pub struct GetShopping;
+//!
+//! impl OperationShape for GetShopping {
+//!     const NAME: &'static str = "GetShopping";
+//!
+//!     type Input = CartIdentifier;
+//!     type Output = ShoppingCart;
+//!     type Error = GetShoppingError;
+//! }
+//! ```
+//!
+//! The behavior of a Smithy operation is encoded by an [`Operation`]. The [`OperationShape`] ZSTs can be used to
+//! construct specific operations using [`OperationShapeExt::from_handler`] and [`OperationShapeExt::from_service`].
+//! The [from_handler](OperationShapeExt::from_handler) constructor takes a [`Handler`] whereas the
+//! [from_service](OperationShapeExt::from_service) takes a [`OperationService`]. Both traits serve a similar purpose -
+//! they provide a common interface over a class of structures.
+//!
+//! ## [`Handler`]
+//!
+//! The [`Handler`] trait is implemented by all closures which accept [`OperationShape::Input`] as their first
+//! argument, the remaining arguments implement [`FromParts`](crate::request::FromParts), and return either
+//! [`OperationShape::Output`] when [`OperationShape::Error`] is [`Infallible`](std::convert::Infallible) or
+//! [`Result`]<[`OperationShape::Output`],[`OperationShape::Error`]>. The following are examples of closures which
+//! implement [`Handler`]:
+//!
+//! ```rust
+//! # use aws_smithy_http_server::Extension;
+//! # pub struct CartIdentifier;
+//! # pub struct ShoppingCart;
+//! # pub enum GetShoppingError {}
+//! # pub struct Context;
+//! # pub struct ExtraContext;
+//! // Simple handler where no error is given.
+//! async fn handler_a(input: CartIdentifier) -> ShoppingCart {
+//!     todo!()
+//! }
+//!
+//! // Handler with an extension where no error is given.
+//! async fn handler_b(input: CartIdentifier, ext: Extension<Context>) -> ShoppingCart {
+//!     todo!()
+//! }
+//!
+//! // More than one extension can be provided.
```

Comment:
Note for the future: I wouldn't advertise this when we get to write the public docs, our only use case so far is supporting Lambda `Context`.